statefun-sdk-go/v3/pkg/statefun/context.go

```diff
@@ -0,0 +1,170 @@
+// and can be used to ensure any spawned go routines do not outlive the current function
+// invocation.
+type Context interface {
+	context.Context
+
+	// Self is the current invoked function instance's Address.
+	Self() Address
+
+	// Caller is the caller function instance's Address, if applicable. This is nil
+	// if the message was sent to this function via an ingress.
+	Caller() *Address
+
+	// Send forwards out a MessageBuilder to another function.
+	Send(message MessageBuilder)
+
+	// SendAfter forwards out a MessageBuilder to another function, after a specified time.Duration delay.
+	SendAfter(delay time.Duration, message MessageBuilder)
+
+	// SendAfterWithCancellationToken forwards out a MessageBuilder to another function,
+	// after a specified time.Duration delay. The message is tagged with a non-empty,
+	//unique token to attach to this message, to be used for message cancellation
+	SendAfterWithCancellationToken(delay time.Duration, token string, message MessageBuilder)
+
+	// CancelDelayedMessage cancels a delayed message (a message that was send via SendAfterWithCancellationToken).
+	// NOTE: this is a best-effort operation, since the message might have been already delivered.
+	// If the message was delivered, this is a no-op operation.
+	CancelDelayedMessage(token string)
+
+	// SendEgress forwards out an EgressBuilder to an egress.
+	SendEgress(egress EgressBuilder)
+
+	// Storage returns the AddressScopedStorage, providing access to stored values scoped to the
+	// current invoked function instance's Address (which is obtainable using Self()).
+	Storage() AddressScopedStorage
+}
+
+type statefunContext struct {
+	sync.Mutex
+	context.Context
+	self     Address
+	caller   *Address
+	storage  *storage
+	response *protocol.FromFunction_InvocationResponse
+}
+
+func (s *statefunContext) Storage() AddressScopedStorage {
+	return s.storage
+}
+
+func (s *statefunContext) Self() Address {
+	return s.self
+}
+
+func (s *statefunContext) Caller() *Address {
+	return s.caller
+}
+
+func (s *statefunContext) Send(message MessageBuilder) {
+	msg, err := message.ToMessage()
+
+	if err != nil {
+		panic(err)
+	}
+
+	invocation := &protocol.FromFunction_Invocation{
+		Target:   msg.target,
+		Argument: msg.typedValue,
+	}
+
+	s.Lock()
+	s.response.OutgoingMessages = append(s.response.OutgoingMessages, invocation)
+	s.Unlock()
+}
+
+func (s *statefunContext) SendAfter(delay time.Duration, message MessageBuilder) {
+	msg, err := message.ToMessage()
+
+	if err != nil {
+		panic(err)
+	}
+
+	invocation := &protocol.FromFunction_DelayedInvocation{
+		Target:    msg.target,
+		Argument:  msg.typedValue,
+		DelayInMs: delay.Milliseconds(),
+	}
+
+	s.Lock()
+	s.response.DelayedInvocations = append(s.response.DelayedInvocations, invocation)
+	s.Unlock()
+}
+
+func (s *statefunContext) SendAfterWithCancellationToken(delay time.Duration, token string, message MessageBuilder) {
+	if len(token) == 0 {
+		panic("cancellation token cannot be empty")
+	}
```

Comment:
`token` is settable by the user, right? Could they handle this error case?

Reply:
Not really, but I think I can use a type alias to more eagerly validate the token.